ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupV2.java

```diff
@@ -0,0 +1,520 @@
+    try {
+      if (includedValidators.isSuperSetOf(attestation.bits())) {
+        return false;
+      }
+    } finally {
+      readLock.unlock();
+    }
+
+    if (committeeShufflingSeedRef.get().isEmpty()) {
+      if (committeeShufflingSeed.isPresent()) {
+        // Attempt to atomically set the value *only if* it's currently Optional.empty()
+        // This guarantees only one thread succeeds in setting the value.
+        committeeShufflingSeedRef.compareAndSet(Optional.empty(), committeeShufflingSeed);
+        // We don't need to check the return value of compareAndSet;
+        // if it fails, it means another thread already set it, which is fine.
+      }
+    }
+
+    if (attestation.isSingleAttestation()) {
+      final Set<PooledAttestation> singleAttestations =
+          singleAttestationsByCommitteeIndex.computeIfAbsent(
+              attestation.bits().getFirstCommitteeIndex(), __ -> ConcurrentHashMap.newKeySet());
+      return singleAttestations.add(attestation);
+    }
+
+    final Set<PooledAttestation> attestations =
+        attestationsByValidatorCount.computeIfAbsent(
+            attestation.bits().getBitCount(), __ -> ConcurrentHashMap.newKeySet());
+
+    // .add() on the ConcurrentHashMap.KeySetView is thread-safe
+    final boolean added = attestations.add(attestation);
+
+    if (earlyDropSingleAttestations && added && attestation.bits().requiresCommitteeBits()) {
+
+      attestation
+          .bits()
+          .streamCommitteeIndices()
+          .forEach(
+              committeeIndex -> {
+                final Set<PooledAttestation> singleAttestations =
+                    singleAttestationsByCommitteeIndex.get(committeeIndex);
+                if (singleAttestations != null) {
+                  singleAttestations.removeIf(sa -> attestation.bits().isSuperSetOf(sa.bits()));
+                }
+              });
+    }
+
+    return added;
+  }
+
+  /**
+   * Iterates through the aggregation of attestations in this group. The iterator attempts to create
+   * the minimum number of attestations that include all attestations in the group.
+   *
+   * <p>committeeIndex is an optional parameter that enables aggregation over a specified committee
+   * (applies to Electra only)
+   *
+   * <p>While it is guaranteed that every validator from an attestation in this group is included in
+   * an aggregate produced by this iterator, there is no guarantee that the added attestation
+   * instances themselves will be included.
+   *
+   * @return an iterator including attestations for every validator included in this group.
+   */
+  private Iterator<PooledAttestation> createAggregatingIterator(
+      final long timeLimitNanos,
+      final Supplier<Stream<PooledAttestation>> candidatesStreamSupplier) {
+    readLock.lock();
+    try {
+      // Capture a copy of includedValidators under lock for the iterator's isolated use
+      AttestationBits includedValidatorsCopy = this.includedValidators.copy();
```

Comment:
final

Reply:
also, we only need the readLock for this copy, dont need for the reduce that unlock right down.

Reply:
ahh yes good catch